### ISSUE TITLE
IF Conditionals

### ISSUE THREAD

**Issue by dwasyl** [iulica/docx-mailmerge#22]
Having had some trouble tracking down how to make `IF` statements work but then eventually noticed the docs:
```
{ IF "{ MERGEFIELD reason }" <> "" "Reason: { MERGEFIELD reason }" }
```
_If the fields are nested inside other fields, the outer fields need to be updated in Word. This can be done by selecting everything (CTRL-a) and then update the fields (F9). There is a way to force the Word to update fields automatically when opening the document. docx-mailmerge can set this setting when saving the document. You can configure this feature by using the auto_update_fields_on_open parameter. The value always will set the setting regardless if needed or not and the value auto will only set it when necessary (when nested fields exist). The default value no will not activate this setting._

Based on that, I've been using the `auto_update_fields_on_open="auto"` setting which works fine, but requires opening the document and saving. In some of my use cases it would be handy to generate a complete Word document and send that the to end-user or another, but when the field needs updating it seems a bit complicated.

Would it be possible for `docx-mailmerge` to handle the conditional statements as part of the processing, or is that logic too complicated/unavailable?

**Comment by iulica** [comment]
It's in the TODO list of another project docx-mergefields (see the documentation). At the moment only the INCLUDEPICTURE is implemented but others are on the way.

The IF is outside the scope of this project, however.

The difficulty is, that the IF field can span bigger chunks of data, like multiple paragraphs and maybe even sections. That makes it much more difficult to implement.

**Comment by augustz** [comment]
I've got the same need. We normally use if's around paragraphs that are optional. If client is marked as wanting a lawn blown clean with electric blowers for example, there is another paragraph that drops in. We also use it in the fee sections. If client wants XXX, then a line and price show up in the fee section.

**Comment by nappex** [comment]
Could be helpful to use [jinja](https://jinja.palletsprojects.com/en/latest/) to implement conditions?

**Comment by iulica** [comment]
Could you give an example ? 
I think there are other template systems to produce pdf docs that use jinja. Here we focus on using Microsoft Word compatibilty, with as few extensions as possible.

**Comment by nappex** [comment]
> Could you give an example ? 
> 
> I think there are other template systems to produce pdf docs that use jinja. Here we focus on using Microsoft Word compatibilty, with as few extensions as possible.

Unforunately I could not I had just idea that docx is just xml and jinja condition pattern could be found there same as it found in html. I have not a lot experience if it is possible. just quick idea

**Comment by iulica** [comment]
well, xml is just the underlying format and not something that you can control from inside the word document. I would say that this is not easily adapted.

On the other hand, as my microsoft office license has changed and I no longer have access to the standalone apps, I would need this feature implemented for myself as well so I will probably add it to the mergefields project in the next months. So stay alert. :)

**Comment by iulica** [comment]
I just commited support for the IF fields. You can test by installing the docx-mailmerge2 directly from github.

Please test it and let me know if everything works.
To enable the IF support, you have to add the enable_experimental=True to the MailMerge object.
Also, make sure you have a space before and after the operator, as described in the docs.

**Comment by iulica** [comment]
Well, I added the release 0.9.0 with the changes activated.

**Comment by iulica** [comment]
@ameboide Thanks for the comment. Your solution is no longer needed, when the IF field works as intended, right ? Or am I missing something?

Although mailmerge has some custom features (e.g. merge rows) that do not work in the Microsoft Word, we try to keep this at a minimum.

**Comment by nappex** [comment]
> I just commited support for the IF fields. You can test by installing the docx-mailmerge2 directly from github.
> 
> 
> 
> Please test it and let me know if everything works.
> 
> To enable the IF support, you have to add the enable_experimental=True to the MailMerge object.
> 
> Also, make sure you have a space before and after the operator, as described in the docs.
> 
> 

Thank you for your work and the new feature ❤️, unfortunately I missed the feature when you needed to test it, but I will try and let you know if some problem occured. Thank you and have a nice day.

**Comment by iulica** [comment]
CMD-F9 and write inside the brackets:  MERGEFIELD reason
then select the whole field and copy it
then write after the field:  <> Reason: 
and then paste the field
copy the double quotes from here "
and paste it everywhere such that you get the following:
"{ MERGEFIELD reason }" <> "" "Reason: { MERGEFIELD reason }"
Then add a new line and press CMD-F9 again to make a new field.
In that field write IF then copy everything from before and paste it after the IF
then you'll get something like this
{ IF "{ MERGEFIELD reason }" <> "" "Reason: { MERGEFIELD reason }" }

Select everything and press F9 (to update the field, important otherwise the fields will not be "complete".
Then save and use the following code:

```
from mailmerge import MailMerge

with MailMerge(
    "tests/test_if_reason.docx",
    remove_empty_tables=False,
    auto_update_fields_on_open="no",
    keep_fields="none",
    enable_experimental=True,
) as document:
    print(document.get_merge_fields())
    document.merge_templates(
        [
            {"reason": ""},
            {"reason": "good reason"},
        ],
        separator="page_break",
    )
    document.write("tests/output/test_if_reason.docx")

```

**Comment by nappex** [comment]
I use your file as template and it does not work for me.

- `test.docx` the template
- `test_if_reason.docx` the output file

my code
```python
with MailMerge(
    "demo/test.docx",
    remove_empty_tables=False,
    auto_update_fields_on_open="no",
    keep_fields="none",
    enable_experimental=True,
) as document:
    print(document.get_merge_fields())
    document.merge_templates(
        [
            {"reason": ""},
            {"reason": "good reason"},
        ],
        separator="page_break",
    )
    document.write("demo/test_if_reason.docx")
```

[test_if_reason.docx](https://github.com/user-attachments/files/23402412/test_if_reason.docx)
[test.docx](https://github.com/user-attachments/files/23402413/test.docx)

**Comment by nappex** [comment]
Yes it is working but I Have a problem that there are old IF conditions which I do not see but they are evaluating also and added their result also (I have three result of IF conditions in output file Concated together). DO you know how could I show all evaluation in docx file, which can affect the result?
Anyway it is not a problem of your code, it will be somethign with word.

**Comment by nappex** [comment]
Yes sorry my explanation was not good.

First time when I tried to find out how to create `IF` condition in word I created this line
`{ IF "{ MERGEFIELD reason }" <> "" "AA" "BB" }`
With `CMD-F9`, but according description noted (the statement logic) when you clicked in word document to `Insert -> Field -> If`.
The description is `Field codes: IF Expression1 Operator Expression2 TrueText FalseText`.

<img width="395" height="464" alt="Image" src="https://github.com/user-attachments/assets/d496dbe3-0d2c-4c11-bd33-69114fa098f0" />

So I understand that the `expression1` is `"{ MERGEFIELD reason }"`, the `Operator` is `<>` and so on.
Then I did not select the line and I didnot press `F9`. I just saved the docx file. When I reopen document again the whole expression disappeared and python code did not work for this expression of `IF`.
Currently when the python code started to work, it is obvious that the disappeared expression is still there because it is adding `AA` or `BB` depends of the value of `reason`, but I do not see the expression anywhere in the docx document. So I can not delete it or edit it.

I hope the explanation is better now. Let me know.
